src/Stratis.Bitcoin.Features.PoA/SlotsManager.cs

```diff
@@ -29,54 +25,79 @@ public class SlotsManager : ISlotsManager
                 throw new NotAFederationMemberException();
 
-            List<IFederationMember> federationMembers = this.federationManager.GetFederationMembers();
-
-            // Round length in seconds.
-            uint roundTime = (uint)this.GetRoundLength(federationMembers.Count).TotalSeconds;
+            var roundTime = this.GetRoundLength(federationMembers.Count);
 
-            // Index of a slot that current node can take in each round.
-            uint slotIndex = (uint)federationMembers.FindIndex(x => x.PubKey == this.federationManager.CurrentFederationKey.PubKey);
+            // Determine the index of the miner that mined the last block.
+            IFederationMember lastMiner = this.federationHistory.GetFederationMemberForBlock(tip);
+            List<IFederationMember> oldFederationMembers = this.federationHistory.GetFederationForBlock(tip);
+            int lastMinerIndex = oldFederationMembers.FindIndex(m => m.PubKey == lastMiner.PubKey);
+            if (lastMinerIndex < 0)
+                throw new Exception($"The miner ('{lastMiner.PubKey.ToHex()}') of the block at height {tip.Height} could not be located in federation.");
 
-            // Time when current round started.
-            uint roundStartTimestamp = (currentTime / roundTime) * roundTime;
-            uint nextTimestampForMining = roundStartTimestamp + slotIndex * this.consensusOptions.TargetSpacingSeconds;
+            int index = -1;
 
-            // Check if we have missed our turn for this round.
-            // We still consider ourselves "in a turn" if we are in the first half of the turn and we haven't mined there yet.
-            // This might happen when starting the node for the first time or if there was a problem when mining.
-            if (currentTime > nextTimestampForMining + (this.consensusOptions.TargetSpacingSeconds / 2) // We are closer to the next turn than our own
-                  || this.chainIndexer.Tip.Header.Time == nextTimestampForMining) // We have already mined in that slot
+            // Looking back, find the first member in common between the old and new federation.
+            for (int i = oldFederationMembers.Count; i >= 1 && index < 0; i--)
             {
-                // Get timestamp for next round.
-                nextTimestampForMining = roundStartTimestamp + roundTime + slotIndex * this.consensusOptions.TargetSpacingSeconds;
+                PubKey keyToFind = oldFederationMembers[(i + lastMinerIndex) % oldFederationMembers.Count].PubKey;
+                index = federationMembers.FindIndex(m => m.PubKey == keyToFind);
+            }
+            
+            if (index < 0)
+                throw new Exception("Could not find a member in common between the old and new federation");
+
+            // Found a member that occurs in both old and new federation.
+            // Determine "distance" apart in new federation.
+            int diff = myIndex - index;
+            while (diff < 0)
```

Comment:
Why the `while` here? wont this just be greater than 0 the moment you add `diff` to  `federationMembers.Count`. I'm struggling to understand some of this.

Reply:
I was thinking that if the "new" federation is substantially different in size more than one iteration may be required. Not much difference between 'if' and 'while' here. It just felt safer.